internal/auth/authn/multiauth.go

```diff
@@ -0,0 +1,796 @@
+package authn
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	api "github.com/flightctl/flightctl/api/v1alpha1"
+	"github.com/flightctl/flightctl/internal/auth/common"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+	"github.com/samber/lo"
+	"github.com/sirupsen/logrus"
+)
+
+// Organization assignment type constants
+const (
+	OrganizationAssignmentTypeStatic  = "static"
+	OrganizationAssignmentTypeDynamic = "dynamic"
+	OrganizationAssignmentTypePerUser = "perUser"
+)
+
+// contextKey is a custom type for context keys to avoid collisions
+type contextKey string
+
+// ParsedTokenCtxKey is the context key for the parsed JWT token
+const ParsedTokenCtxKey contextKey = "parsed_token"
+
+// GetParsedTokenFromContext retrieves the parsed JWT token from the context if it exists
+func GetParsedTokenFromContext(ctx context.Context) (jwt.Token, bool) {
+	if token, ok := ctx.Value(ParsedTokenCtxKey).(jwt.Token); ok {
+		return token, true
+	}
+	return nil, false
+}
+
+// TokenType represents the type of JWT authentication token
+type TokenType int
+
+const (
+	TokenTypeOIDC TokenType = iota
+	TokenTypeK8s
+)
+
+// AuthProviderService interface for auth provider operations
+type AuthProviderService interface {
+	ListAuthProviders(ctx context.Context, params api.ListAuthProvidersParams) (*api.AuthProviderList, api.Status)
+	GetAuthProviderByIssuerAndClientId(ctx context.Context, issuer string, clientId string) (*api.AuthProvider, api.Status)
+}
+
+// AuthProviderCacheKey is a composite key for caching auth providers
+type AuthProviderCacheKey struct {
+	Issuer   string
+	ClientId string
+}
+
+// MultiAuth implements authentication using multiple providers with issuer-based routing
+type MultiAuth struct {
+	// Static providers - initialized once at startup, mapped by issuer
+	staticProviders map[string]common.AuthNMiddleware // issuer -> provider mapping
+
+	// Service for dynamic auth providers
+	authProviderService AuthProviderService
+
+	// TLS config for OIDC provider connections
+	tlsConfig *tls.Config
+
+	// Logger for authentication operations
+	log logrus.FieldLogger
+
+	// Dynamic OIDC providers - issuer+clientId -> provider mapping
+	dynamicProviders   map[AuthProviderCacheKey]common.AuthNMiddleware
+	dynamicProvidersMu sync.RWMutex
+
+	// Control for background loader goroutine
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+// NewMultiAuth creates a new MultiAuth instance
+func NewMultiAuth(authProviderService AuthProviderService, tlsConfig *tls.Config, log logrus.FieldLogger) *MultiAuth {
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
creating ctx in constructor feels awkward.  can we use Run(ctx)  pattern or if you need off switch you can use `done <-chan struct{}`

```go

func (s *Shutdown) Close() {
    select {
    case <-s.done:
        return // already closed
    default:
        close(s.done)
    }
}
```